### DISCUSSION TITLE
Handle with cluster (virtual chassis in netbox) not working nicely

### DISCUSSION THREAD

**Discussion by pc86** [bonzo81/netbox-librenms-plugin#19] [General]
Hello
The behavior of the interface sync is not working smoothly with stack.
we have a 2-device stack .
In LibreNMS => it is seen as a single Device 
In NetBox => it is defined as **2** devices linked as virtual chassis, with the 1 device as primary member

All the interface of the stack are displayed in the first member interface (but still belong to the second member) 

The full sync interface create interface of the second member in the first member (make it double: exemple with interface 2/1/10)

![{BDE53A47-7318-4074-9EA4-6620C2D1630E}](https://github.com/user-attachments/assets/4bda4bc7-4ac8-406f-8ac7-fbf5a7ec7a1a)

But I think it might not be very simple to handle this properly.
Best regards

**Comment by bonzo81** [comment]
Hi @pc86,

Yes you are right. At this point the plugin doesn't have the logic to differentiate between the different stack member ports. It is currently on the plugin roadmap to introduce this logic. There is an issue created for this already. See https://github.com/bonzo81/netbox-librenms-plugin/issues/3. 

I hope to add this to the next release.

Thanks

**Comment by pc86** [comment]
This might be difficult, as LibreNMs and Netbox doesn't have the same logic.
For LibreNMS, the stack is a single device, while in Netbox, it is 2 differents devices linked thru a virtual chassis (all the interface are then available thru the first member). The virtual chassis doesn't exist by itself.

**Comment by bonzo81** [comment]
@pc86 I've got something working that should handle virtual chassis logic. 

* New column in table for virtual Chassis (VC) showing the VC member to sync the interface to.
* The plugin will attempt to automatically select the correct member by comparing VC position number to the 1st number in the interface.
* Selecting a new member will trigger a new interface details comparison against the newly selected device.
* This is handled on either the Virtual Chassis master, or the first member with a primary IP.

Hoping to get a new release next week.

Here's a quick preview

![vc_member_select](https://github.com/user-attachments/assets/87dd9e04-c26c-4dfc-8c5d-453099261ce8)

**Comment by pc86** [comment]
Great news !
Just let me know for a test

Le ven. 8 nov. 2024 à 19:32, Andy Norwood ***@***.***> a
écrit :

**Comment by bonzo81** [comment]
Hi @pc86,

v0.2.7 just released with virtual chassis support. Let me know how you get on, and please feedback any issues.

**Comment by pc86** [comment]
Hi,
Great and fast job :)

Sync is looking good - but I made a very short tests

*Few thoughts*
1) Having this information on the second cluster member is very useful
[image: {8375D911-BDA7-4550-AE48-B321C8CCB68E}.png]

2)  ability to filter interfaces could be helpfull (I have about 400
interface / lag / vlan on this cluster) - but I don't know the dev cost.
3) forcing the cluster member for a selection of interface too

*General behavior*
1) For some unsync interfaces, the second member is automatically selected,
- but only very few
[image: image.png]

2) Assigning the cluster member for a set of interface could be nice -
again I don't know the dev cost.

*UI*
The display of select box of the cluster member when you select it change
between a navigator is in full screen mode vs normal windows.
(I know I might not be clear - it's a ui minor problem)

Best Regards
Philippe

Le lun. 11 nov. 2024 à 17:11, Andy Norwood ***@***.***> a
écrit :

**Comment by bonzo81** [comment]
Hi,

Thanks for the feedback, all good helpful ideas.

Could you upload your images again as they don't appear here. I'm interested to know more about if the plugin isn't selecting the right member.

I'm unsure about the UI issue you reported. Could you share an image of it? I can't replicate anything similar here.

Thanks.

**Comment by pc86** [comment]
I am out of my office today. I will do it tomorrow.
I will also try to make a short animation about the ui as it is hard to
explain.
Best regards


Le mar. 12 nov. 2024 à 13:10, Andy Norwood ***@***.***> a
écrit :

**Comment by pc86** [comment]
Hello
Sorry for the delay, it is easier for me to handle this the weekend :)

I tested with a "fresh cluster" - aka not yet sync with librenms

Most of the interfaces of the second member (XXXX**4**) are stuck to the first member(XXXX**3**).

![image](https://github.com/user-attachments/assets/4030b032-7063-40fe-87c0-96e500afb3ad)

I have 4 type of interface in these switches
- M-GigabitEthernet -> first member
- Ten-GigabitEthernet -> always first member
- Twenty-FiveGigE -> always first member
- FortyGigE / HundredGigE -> always well assigned (second member) - there the same interace - it depend on the transceiver in 

All the vlan (Vlan-interfaceXXX) and the lag (Bridge-AggregationXX) are on the first member

I haven't yet sync this cluster, so if you want to test a version, just let me know - the only point is that I am using the netbocker docker version, I don't know how to the dev version of a plugin... yet

**Reply by bonzo81** [reply]
Thanks for spending time to reply.

The port data from LibreNMS does not include which VC member the interface is attached to. Most likely because the SNMP data from the device doesn't include it either.

So I built some logic into the plugin to assign an interface to a VC member, and it is quite simple at the moment. Essentially, it tries to match the first number in the interface (x) to a VC member with the same VC position number (x) . It only does this for interfaces that have 'ethernet' in their ifType value. So Vlan, Stack, loopback etc will not be attempt this logic because their ifType does not contain 'ethernet'. This was based on the kit I have available in my own environment.

Here is a flow diagram of the logic at the moment.
![VC interface logic](https://github.com/user-attachments/assets/72e38128-9040-4941-8e60-04ae2190d9cb)

What are the interface types displayed before applying the interface mappings? 

If you can see a possible improvement to the logic for your environment please suggest it. I think it might not be possible to get the logic right for every eventuality. That's why I created the drop down!

I will be adding the ability to change multiple interfaces at once very soon, so hopefully that will help too.

Thanks for your feedback so far.

**Comment by pc86** [comment]
I forgot a functionnality you may consider
We are using SNMPv3, adding a device from Netbox to LibreNMS using SNMPv3 would be great (but again, I don't know the dev cost).

**Reply by bonzo81** [reply]
Thanks for this. I've created an issue for it #29

**Comment by pc86** [comment]
About the UI
I made capture before/after I selet the cluster member, 
the listbox enlarges when opened as a consequence the layout changes

![{F9785E52-0356-441E-A571-45E1FFB75EC6}](https://github.com/user-attachments/assets/a0aa1780-2494-43a7-b4a0-7ad465ca5e50)

**Reply by bonzo81** [reply]
Created an issue for this too #28
